src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+            arguments = AddExpressionArgumentToArgumentList(arguments, expression.WithoutTrailingTrivia());
+            var memberName = methodSymbol.IsGenericMethod
+                ? generator.GenericName(methodSymbol.Name, methodSymbol.TypeArguments)
+                : generator.IdentifierName(methodSymbol.Name);
+
+            var invocation = generator.InvocationExpression(memberName, arguments);
+            List<SyntaxNode> invocationReturn;
+
+            if (!methodSymbol.ReturnsVoid)
+            {
+                invocationReturn = new List<SyntaxNode>
+                {
+                    generator.ReturnStatement(invocation)
+                };
+            }
+            else
+            {
+                invocationReturn = new List<SyntaxNode>
+                {
+                    invocation
+                };
+            }
+
+            var newMethod = CodeGenerationSymbolFactory.CreateMethodSymbol(methodSymbol);
+
+            var methodDeclaration = generator.MethodDeclaration(newMethod, invocationReturn);
+
+            editor.InsertBefore(methodExpression, methodDeclaration);
+        }
+
+        public static void UpdateExpressionInOriginalFunction(SemanticDocument semanticDocument,
+                                                              TExpressionSyntax expression, SyntaxNode scope,
+                                                              string parameterName, SyntaxEditor editor,
+                                                              bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(semanticDocument.Document);
+            var matches = FindMatches(semanticDocument, expression, semanticDocument, scope, allOccurrences, cancellationToken);
+            var parameterNameSyntax = (TIdentifierNameSyntax)generator.IdentifierName(parameterName);
+            // Parenthesize the variable, and go and replace anything we find with it.
+            // NOTE: we do not want elastic trivia as we want to just replace the existing code 
+            // as is, while preserving the trivia there.  We do not want to update it.
+
+            var replacement = generator.AddParentheses(parameterNameSyntax, includeElasticTrivia: false)
+                                         .WithAdditionalAnnotations(Formatter.Annotation);
+
+            var syntaxFacts = semanticDocument.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            foreach (var match in matches)
+            {
+                if (match.Equals(expression))
+                {
+                    if (IsExpressionRemovable(semanticDocument, expression))
+                    {
+                        var localDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => syntaxFacts.IsLocalDeclarationStatement(node));
+                        if (localDeclaration is not null)
```

Comment:
how coudl this be null given the if check above.
this sort of check is bery problematic as it means we end up doing nohting even though we're in this codepath because we checked to make sure we should do something.

Reply:
I don't think it can be null, but I get warnings that localDeclaration could be null. Is there something I can do to fix that without doing a nnull check?

Reply:
if it must be null because of some Roslyn invariant (that we can't encode in the typesystem) then use Contract.ThrowIfNull.

Reply:
ShouldRemoveVariableDeclaratorContainingExpression should just return the LocalDeclStatement that we would want to remove.